### PR TITLE
Random Beacon v2 relay entry timeout and slashing

### PR DESCRIPTION
**Pull request by lukasz-zimnoch**
Depends on: #2637 

This pull request introduces timeout and slashing capabilities to the relay entry feature. More details about specific changes can be found below.

### Relay entry timeout reporting

If the relay entry hard timeout (`64 * 10 + 5760`) exceeds, the `submitRelayEntry` function starts reverting upon calls. This PR introduces a way to report such a timeout using the `reportRelayEntryTimeout` function. That function closes the current relay request, slashes the whole stake of all members in the group, terminates the group, and retries the request by creating a new one but without requiring the request fee.

### Punishment for exceeding the soft timeout

If the relay entry soft timeout (`64 * 10`) exceeds, there is still a possibility to call `submitRelayEntry` and fulfill the request. However, members start bleeding linearly and a portion of their stakes is getting slashed, depending on the moment the relay entry gets submitted. The exact slashing amount is computed as: 
```
slashingAmount = relayEntrySubmissionFailureSlashingAmount * slashingFactor
```
 where 
```
slashingFactor = (block.number - softTimeoutBlock) * 100% / relayEntryHardTimeout
```
For example, if 
```
relayEntrySubmissionFailureSlashingAmount = 1000
relayEntryHardTimeout = 5760
```
and the relay entry has been submitted `2880` after the soft timeout block, the actual slashing amount can be computed as follows: 
```
slashingFactor = 2880 * 100% / 5760 = 50%
slashingAmount = 1000 * 50% = 500
```

### Punishment of members who not submit the relay entry on their turn

In #2637 we introduced the submission eligibility queue which determines the order group members should preserve while submitting the relay entry result. The first eligible member index is computed as `(entry % groupSize) + 1`.
According to the RFC, all members who did not submit the relay entry on their turn, should be kicked out from the sortition pool for a given period of time. In effect, they don't receive rewards. This PR introduces such a mechanism of computing the list of punished members. Let's suppose we have a group of `8` members, the first eligible member index is `6` and the result is submitted by member `3`:
```
1, 2, 3 (submitter), 4, 5, 6 (first eligible), 7, 8
```
In that case, the list of punished members is: `6, 7, 8, 1, 2`. Worth noting that the actual mechanism of kicking out from the sortition pool is not implemented within this PR and will be addressed in a follow-up PR: https://github.com/keep-network/keep-core/pull/2671.

### CODE REVIEW

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 1]
```diff
@@ -309,4 +393,76 @@ library Relay {
...
+        internal
+        view
+        returns (uint256)
+    {
+        uint256 softTimeoutBlock = self.currentRequest.startBlock +
+            (_groupSize * self.relayEntrySubmissionEligibilityDelay);
+
+        if (block.number > softTimeoutBlock) {
+            uint256 submissionDelay = block.number - softTimeoutBlock;
+            uint256 slashingFactor = (submissionDelay * 1e18) /
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/random-beacon/contracts/libraries/Relay.sol:461]: The slashing amount needs to be reasonable amount so that it makes sense for someone to process the slashing queue. Please see https://github.com/threshold-network/solidity-contracts/pull/7 to see how slashing is implemented there. I think we can relay on the Governance to set a reasonable slashing value but at the same time, we should ensure `.slash` is not called when `slashingAmount = 0`. This could be the case for the first few weeks - something we will have to decide on.

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 2]
```diff
@@ -160,12 +200,25 @@ library Relay {
...
+        self.sortitionPool.removeOperators(punishedMembers);
+
+        // If the soft timeout has been exceeded apply stake slashing for
+        // all group members. Note that `getSlashingFactor` returns the
+        // factor multiplied by 1e18 to avoid precision loss. In that case
+        // the final result needs to be divided by 1e18.
+        uint256 slashingAmount = (getSlashingFactor(self, groupSize) *
+            self.relayEntrySubmissionFailureSlashingAmount) / 1e18;
+        // slither-disable-next-line reentrancy-events
+        self.staking.slash(slashingAmount, group.members);
```
Comment (pdyraga) [inline solidity/random-beacon/contracts/libraries/Relay.sol:221]: If we decide to slash only in `RandomBeacon` (which IMO would be a nice pattern), we could return slashing amount from this function and call `staking.slash` in `RandomBeacon` only if this amount is non-zero.

**Review by pdyraga — COMMENTED** [review]
Was not reviewing tests, leaving it for later.

Before we merge we need to confirm what is the cost of submitting relay entry for various edge cases:
- no operators removed from the pool
- a lot of operators removed from the pool
- soft deadline reached

If we decide not to work against a real sortition pool in this PR, we will have to repeat these checks later, once we integrate sortition pool.

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts [hunk 3]
```diff
@@ -154,27 +178,152 @@ describe("RandomBeacon - Relay", () => {
...
+                      .connect(member16)
+                      .submitRelayEntry(
+                        firstEligibleMemberIndex,
+                        blsData.groupSignature
+                      )
+
+                    receipt = await tx.wait()
+                  })
+
+                  it("should not remove any members from the sortition pool", async () => {
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts:289]: True, members 4 and 5 can be considered lucky in this scenario but there is a risk of not being that lucky. I could not come up with anything better. Marking members as ineligible to submit once they turn passed can have other serious consequences - if the submission window is too short for the current mainnet state, the entire group can easily reach a soft timeout.

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 4]
```diff
@@ -206,6 +260,37 @@ library Relay {
...
+    }
+
+    /// @notice Reports a relay entry timeout.
+    /// @param group Group data.
+    function reportEntryTimeout(Data storage self, Groups.Group memory group)
+        internal
+    {
+        require(hasRequestTimedOut(self), "Relay request did not time out");
+
+        // slither-disable-next-line reentrancy-events
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol]: See https://github.com/keep-network/keep-core/pull/2664/commits/5ade285f4392b4c579d6b4afdbe7dc2e9c7e8bad

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 5]
```diff
@@ -206,6 +260,37 @@ library Relay {
...
+        self.staking.slash(
+            self.relayEntrySubmissionFailureSlashingAmount,
+            group.members
+        );
+
+        delete self.currentRequest;
+
+        emit RelayEntryTimedOut(self.requestCount);
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol]: See https://github.com/keep-network/keep-core/pull/2664/commits/5ade285f4392b4c579d6b4afdbe7dc2e9c7e8bad

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 6]
```diff
@@ -309,4 +393,76 @@ library Relay {
...
+    ) internal view returns (address[] memory) {
+        uint256 punishedMembersCount = _submitterIndex >= _firstEligibleIndex
+            ? _submitterIndex - _firstEligibleIndex
+            : _groupSize - (_firstEligibleIndex - _submitterIndex);
+
+        address[] memory punishedMembers = new address[](punishedMembersCount);
+
+        for (uint256 i = 0; i < punishedMembersCount; i++) {
+            uint256 memberIndex = _firstEligibleIndex + i;
+            memberIndex = memberIndex > _groupSize
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol]: Right, see https://github.com/keep-network/keep-core/pull/2664/commits/fb96c8871514f8966a06a9e2513599b535f81b95

File: solidity/random-beacon/contracts/libraries/Relay.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol:470]: Done in https://github.com/keep-network/keep-core/pull/2664/commits/7f1dab2e14327d3364ab400f29513bc1dc535037

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts [hunk 7]
```diff
@@ -247,13 +396,65 @@ describe("RandomBeacon - Relay", () => {
...
+            .submitRelayEntry(
+              firstEligibleMemberIndex,
+              blsData.nextGroupSignature
+            )
         ).to.be.revertedWith("No relay request in progress")
       })
     })
   })
 
+  describe("reportRelayEntryTimeout", () => {
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts:413]: I was thinking about it once again and we will leave it as is. An explicit assertion wouldn't make sense as long as the timeout reporter doesn't make an explicit T token allowance. On the other hand, if the reporter makes such an allowance, this test can suggest the reporter needs to pay a fee in T tokens which are not true.

File: solidity/random-beacon/contracts/RandomBeacon.sol [hunk 8]
```diff
@@ -32,6 +32,18 @@ interface ISortitionPool {
     function isOperatorInPool(address operator) external view returns (bool);
 
     function isOperatorEligible(address operator) external view returns (bool);
+
+    // TODO: Temporary function for testing only. Not sure how the removal
+    //       mechanism will work at the moment.
+    function removeOperators(address[] memory operators) external;
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: Yeah, that's clear now. This TODO comes from times where I wasn't sure about that. The signature of this method has been changed as part of https://github.com/keep-network/keep-core/pull/2664/commits/c7cf1edf952ebb948e413e5709b6c050db041600

Regarding tests against a real pool, I agree. However, I'd love to introduce them after this PR and #2671 land on `main`. This will make things far easier given the fact the whole beacon codebase is under heavy development and things change very quickly which creates conflicts. Also, I think making gas optimizations on `main` code would be easier than making them on stacked PRs.

Regarding TODO about `keep-network/sortition-pools`, added it within: https://github.com/keep-network/keep-core/pull/2664/commits/b89888d0e34c8a85acc5adb7180c3bbfdceb082e

File: solidity/random-beacon/contracts/RandomBeacon.sol [hunk 9]
```diff
@@ -32,6 +32,18 @@ interface ISortitionPool {
...
+}
+
+/// @title Staking contract interface
+/// @notice This is an interface with just a few function signatures of the
+///         Staking contract, which is available at
+///         https://github.com/threshold-network/solidity-contracts/blob/main/contracts/staking/IStaking.sol
+interface IStaking {
+    function slash(uint256 amount, address[] memory operators) external;
 }
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:58]: Left a TODO: https://github.com/keep-network/keep-core/pull/2664/commits/8a70c0aca36fd4ceac8bf8d27312cdda1d0678fd

File: solidity/random-beacon/contracts/RandomBeacon.sol [hunk 10]
```diff
@@ -173,13 +174,19 @@ contract RandomBeacon is Ownable {
 
     event RelayEntrySubmitted(uint256 indexed requestId, bytes entry);
 
+    event RelayEntryTimedOut(uint256 indexed requestId);
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:188]: The report timeout action will cause a retry if of the relay request thus will emit a new `RelayEntryRequested` with the same request ID but new `groupId`. I don't think we need to add it here.

File: solidity/random-beacon/contracts/RandomBeacon.sol [hunk 11]
```diff
@@ -512,6 +533,27 @@ contract RandomBeacon is Ownable {
...
+        uint64 groupId = relay.currentRequest.groupId;
+        relay.reportEntryTimeout(groups.getGroup(groupId));
+
+        // TODO: Once implemented, invoke:
+        // terminateGroup(groupId);
+
+        // In case we retry the timed out request, we can't require the
+        // the request fee to be payed.
+        if (groups.numberOfActiveGroups() > 0) {
+            _requestRelayEntry(callback.callbackContract, false);
```
[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/random-beacon/contracts/RandomBeacon.sol:560]: We are talking about `RandomBeacon.reportRelayEntryTimeout`, right? If so, I'd do just if-else and in else, I'd call another function of `Relay`, say `cleanupOnEntryTimeout`.

File: solidity/random-beacon/contracts/libraries/Relay.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (dimpar) [reply solidity/random-beacon/contracts/libraries/Relay.sol:470]: ok, looks good

File: solidity/random-beacon/contracts/RandomBeacon.sol [hunk 12]
```diff
@@ -512,6 +533,27 @@ contract RandomBeacon is Ownable {
         callback.executeCallback(uint256(keccak256(entry)), callbackGasLimit);
     }
 
+    /// @notice Reports a relay entry timeout.
+    function reportRelayEntryTimeout() external {
+        uint64 groupId = relay.currentRequest.groupId;
+        relay.reportEntryTimeout(groups.getGroup(groupId));
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: Good idea. Done as part of https://github.com/keep-network/keep-core/pull/2664/commits/e357f7d1b0266e7449edfcb7bc69e7af58dda7bf

File: solidity/random-beacon/contracts/RandomBeacon.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol]: Ok. Please see https://github.com/keep-network/keep-core/pull/2664/commits/e357f7d1b0266e7449edfcb7bc69e7af58dda7bf

File: solidity/random-beacon/contracts/RandomBeacon.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:189]: In case of request retry we emit a new `RelayEntryRequested` event (see `retryOnEntryTimeout` introduced by https://github.com/keep-network/keep-core/pull/2664/commits/e357f7d1b0266e7449edfcb7bc69e7af58dda7bf) so the new group is aware. I don't think we need to put `groupId` here.

File: solidity/random-beacon/contracts/test/SortitionPoolStub.sol [hunk 13]
```diff
@@ -35,4 +37,15 @@ contract SortitionPoolStub is ISortitionPool {
...
+    function removeOperators(address[] memory _operators) external override {
+        for (uint256 i = 0; i < _operators.length; i++) {
+            delete operators[_operators[i]];
+            delete eligibleOperators[_operators[i]];
+        }
+
+        if (_operators.length > 0) {
+            emit OperatorsRemoved(_operators);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/test/SortitionPoolStub.sol:75]: Yeah, that's clear (see also https://github.com/keep-network/keep-core/pull/2664#discussion_r740973298)

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 14]
```diff
@@ -160,12 +200,25 @@ library Relay {
...
 
-        // TODO: Use submitterIndex, firstEligibleIndex and lastEligibleIndex
-        //       to prepare an array of addresses which should be kicked from
-        //       the sortition pool for 2 weeks.
-
-        // TODO: If soft timeout has elapsed, take bleeding into account
-        //       and slash all members appropriately.
+        // Get the list of members addresses which should be punished due to
+        // not submitting the entry on their turn.
+        address[] memory punishedMembers = getPunishedMembers(
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol]: Good point, see https://github.com/keep-network/keep-core/pull/2664/commits/c37e3bca3bfc2f7550857735015d203720bbf6d6

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 15]
```diff
@@ -160,12 +200,25 @@ library Relay {
...
+            self,
+            submitterIndex,
+            firstEligibleIndex,
+            group,
+            groupSize
+        );
+        self.sortitionPool.removeOperators(punishedMembers);
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol]: Yep, already added it to `ISortitionPool`. Though, it's not required here as we also have `removeOperators(uint32[] calldata ids)`

File: solidity/random-beacon/contracts/libraries/Relay.sol (hunk 2 above)
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol]: Sure. I will handle it in follow-up https://github.com/keep-network/keep-core/pull/2671 where I do the same for `inactiveMembers`. This way we don't block here with tests and we don't introduce new conflicts.

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 16]
```diff
@@ -160,12 +200,25 @@ library Relay {
...
-        // TODO: If soft timeout has elapsed, take bleeding into account
-        //       and slash all members appropriately.
+        // Get the list of members addresses which should be punished due to
+        // not submitting the entry on their turn.
+        address[] memory punishedMembers = getPunishedMembers(
+            self,
+            submitterIndex,
+            firstEligibleIndex,
+            group,
+            groupSize
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol]: Yep, I have not taken into account that the actual group size can be <64 due to DKG misbehavior. Thanks for catching this! Fixed in https://github.com/keep-network/keep-core/pull/2664/commits/3070a88f1dde19cbf821009aed739051bb56ed5b. Please watch this commit carefully as it contains some other changes described below.

Regarding the other changes from the commit, I have realized we need to use the actual (not the ideal) group size for eligibility calculations. My reasoning is: if the group doesn't have an ideal size, we should make their members eligible as usual so we must use the actual size, not the ideal one. If we use the ideal size during eligibility calculations, we will introduce eligibility delays for members whose indexes are placed after the misbehaved ones.

On the other hand, I think we should use the ideal group size for calculations related to the soft timeout (`hasRequestTimedOut` and `getSlashingFactor`). The reasoning here: if a group doesn't have the ideal size, we shouldn't punish well-behaving members by giving a shorter soft timeout for the entire group. 

Is that correct?

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 17]
```diff
@@ -309,4 +393,76 @@ library Relay {
...
+    /// @param _groupSize _groupSize Group size.
+    /// @return An array of members addresses which should be punished due to
+    ///         not submitting a relay entry on their turn.
+    function getPunishedMembers(
+        /* solhint-disable-next-line no-unused-vars */
+        Data storage self,
+        uint256 _submitterIndex,
+        uint256 _firstEligibleIndex,
+        Groups.Group memory _group,
+        uint256 _groupSize
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol]: See https://github.com/keep-network/keep-core/pull/2664#discussion_r741198684

**Comment by lukasz-zimnoch** [comment]
@pdyraga all comments addressed!

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts [hunk 18]
```diff
@@ -164,27 +185,152 @@ describe("RandomBeacon - Relay", () => {
...
+                      .connect(member16)
+                      .submitRelayEntry(
+                        firstEligibleMemberIndex,
+                        blsData.groupSignature
+                      )
+
+                    receipt = await tx.wait()
+                  })
+
+                  it("should not remove any members from the sortition pool", async () => {
```
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.Relay.test.ts:296]: Can we add another context "when some other than the first eligible member submits after the soft timeout" and ensure some operators are removed from the pool?

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts [hunk 19]
```diff
@@ -164,27 +185,152 @@ describe("RandomBeacon - Relay", () => {
...
+                    await mineBlocks(64 * 10 + 0.75 * 5760 - 1)
+
+                    tx = await randomBeacon
+                      .connect(member16)
+                      .submitRelayEntry(
+                        firstEligibleMemberIndex,
+                        blsData.groupSignature
+                      )
+
+                    receipt = await tx.wait()
```
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.Relay.test.ts:293]: Unused `receipt`!

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 20]
```diff
@@ -178,12 +199,31 @@ library Relay {
...
+            firstEligibleIndex,
+            group.members
+        );
+        self.sortitionPool.removeOperators(inactiveMembers);
+
+        // If the soft timeout has been exceeded apply stake slashing for
+        // all group members. Note that `getSlashingFactor` returns the
+        // factor multiplied by 1e18 to avoid precision loss. In that case
+        // the final result needs to be divided by 1e18.
+        uint256 slashingAmount = (getSlashingFactor(self, idealGroupSize) *
```
Comment (pdyraga) [inline solidity/random-beacon/contracts/libraries/Relay.sol:216]: Using `idealGroupSize` here 👌🏻 💯 Same for `relayEntryTimeout`. Good reasoning!

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts [hunk 21]
```diff
@@ -257,78 +403,332 @@ describe("RandomBeacon - Relay", () => {
...
+            )
+
+            // `groupSize * relayEntrySubmissionEligibilityDelay +
+            // relayEntryHardTimeout`.
+            await mineBlocks(64 * 10 + 5760)
+
+            tx = await randomBeacon.reportRelayEntryTimeout()
+          })
+
+          it("should slash entire stakes of all group members", async () => {
```
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.Relay.test.ts:449]: Nah, not entire stakes! It's a full slashing amount for the timeout but not the entire stake. Let's update this test name to do not scare anyone haha.

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts [hunk 22]
```diff
@@ -257,78 +403,332 @@ describe("RandomBeacon - Relay", () => {
...
+          it("should emit RelayEntryTimedOut event", async () => {
+            await expect(tx)
+              .to.emit(randomBeacon, "RelayEntryTimedOut")
+              .withArgs(1)
+          })
+
+          it("should retry current relay request", async () => {
+            // We expect the same request ID because this is a retry.
+            // Group ID is still `0` because there is only one group
+            // after termination was performed.
```
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.Relay.test.ts:468]: Heads up, this will no longer be true once @dimpar implements group termination. Group ID is equal to the group's position in the groups array, so we will end up with index 1 here.

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts [hunk 23]
```diff
@@ -257,78 +403,332 @@ describe("RandomBeacon - Relay", () => {
...
+
+        beforeEach(async () => {
+          // `groupSize * relayEntrySubmissionEligibilityDelay +
+          // relayEntryHardTimeout`.
+          await mineBlocks(64 * 10 + 5760)
+
+          tx = await randomBeacon.reportRelayEntryTimeout()
+        })
+
+        it("should slash entire stakes of all group members", async () => {
```
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.Relay.test.ts:489]: Same comment about the test name 😅

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts [hunk 24]
```diff
@@ -257,78 +403,332 @@ describe("RandomBeacon - Relay", () => {
...
+        it("should emit RelayEntryTimedOut event", async () => {
+          await expect(tx)
+            .to.emit(randomBeacon, "RelayEntryTimedOut")
+            .withArgs(1)
+        })
+
+        it("should clean up current relay request data", async () => {
+          // TODO: Uncomment those assertions once termination is implemented.
+          // await expect(tx).to.not.emit(randomBeacon, "RelayEntryRequested")
+          // expect(await randomBeacon.isRelayRequestInProgress()).to.be.false
```
Comment (pdyraga) [inline solidity/random-beacon/test/RandomBeacon.Relay.test.ts:508]: Should we uncoment them?

**Review by pdyraga — APPROVED** [review]
Let's address review comments in the next PR that depends on this one.

File: solidity/random-beacon/contracts/RandomBeacon.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:189]: Just to be clear, the `groupId` we are talking about here means "the group ID of the terminated group" not the "group ID of the group selected for retry" right? If so, maybe let's call it `terminatedGroupId` to avoid confusion we had in beacon v1?

[the remark this answers is not in the document]
Comment (pdyraga) [reply solidity/random-beacon/contracts/RandomBeacon.sol:189]: Yep! The one for which the entry timed out. I think `terminatedGroupId` is 👌🏻

[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/RandomBeacon.sol:189]: Addressed in https://github.com/keep-network/keep-core/pull/2671/commits/a8a0882231843158d659a7b75e3b209ccf8ef092 part to of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 25]
```diff
@@ -39,21 +39,28 @@ library Relay {
...
+        uint256 relayEntrySubmissionFailureSlashingAmount;
     }
 
-    /// @notice Size of a group in the threshold relay.
-    uint256 public constant groupSize = 64;
+    /// @notice Ideal size of a group in the threshold relay. A group has
+    ///         an ideal size if all their members behaved properly during
+    ///         group formation. Actual group size can be lower in groups
+    ///         with proven misbehaved members.
+    uint256 public constant idealGroupSize = 64;
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol:63]: Addressed in https://github.com/keep-network/keep-core/pull/2671/commits/2b51ee823eef477f880d66e4389e418c96965f79 being part of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 26]
```diff
@@ -145,11 +165,12 @@ library Relay {
         Groups.Group memory group
     ) internal {
         require(isRequestInProgress(self), "No relay request in progress");
-        // TODO: Add timeout reporting.
         require(!hasRequestTimedOut(self), "Relay request timed out");
 
+        uint256 actualGroupSize = group.members.length;
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol:170]: Addressed in https://github.com/keep-network/keep-core/pull/2671/commits/f447e647fd33344510e503a27d7a3fc2415a11ad being part of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 27]
```diff
@@ -327,4 +409,73 @@ library Relay {
...
+    /// @return An array of members IDs which should be  inactive due
+    ///         to not submitting a relay entry on their turn.
+    function getInactiveMembers(
+        /* solhint-disable-next-line no-unused-vars */
+        Data storage self,
+        uint256 _submitterIndex,
+        uint256 _firstEligibleIndex,
+        uint32[] memory _groupMembers
+    ) internal view returns (uint32[] memory) {
+        uint256 _groupSize = _groupMembers.length;
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol:431]: Addressed in https://github.com/keep-network/keep-core/pull/2671/commits/f447e647fd33344510e503a27d7a3fc2415a11ad being part of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 28]
```diff
@@ -327,4 +409,73 @@ library Relay {
...
+    /// @param _firstEligibleIndex First index of the given eligibility range.
+    /// @param _groupMembers IDs of the group members.
+    /// @return An array of members IDs which should be  inactive due
+    ///         to not submitting a relay entry on their turn.
+    function getInactiveMembers(
+        /* solhint-disable-next-line no-unused-vars */
+        Data storage self,
+        uint256 _submitterIndex,
+        uint256 _firstEligibleIndex,
+        uint32[] memory _groupMembers
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol:429]: Initially, this method accepted `groupSize` as a param that conflicted with `groupSize` library constant. That's no longer the case here so we can get rid of the underscores. Addressed in 
https://github.com/keep-network/keep-core/pull/2671/commits/00c3e2d9d0b01e0e8f4713a5cef1864eb86a188b being part of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/contracts/libraries/Relay.sol [hunk 29]
```diff
@@ -178,12 +199,31 @@ library Relay {
...
+        // If the soft timeout has been exceeded apply stake slashing for
+        // all group members. Note that `getSlashingFactor` returns the
+        // factor multiplied by 1e18 to avoid precision loss. In that case
+        // the final result needs to be divided by 1e18.
+        uint256 slashingAmount = (getSlashingFactor(self, idealGroupSize) *
+            self.relayEntrySubmissionFailureSlashingAmount) / 1e18;
+
+        // TODO: This call will be removed from here in the follow-up PR.
+        if (slashingAmount > 0) {
+            // slither-disable-next-line reentrancy-events
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/contracts/libraries/Relay.sol:221]: Addressed in 
https://github.com/keep-network/keep-core/pull/2671/commits/f3464a249b55067b2a56a4fae82515d6552217e5 being part of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/test/utils/groups.ts [hunk 30]
```diff
@@ -3,7 +3,7 @@ import { signAndSubmitDkgResult } from "./dkg"
 import { constants, params } from "../fixtures"
 import blsData from "../data/bls"
 import { Operator } from "./sortitionpool"
-import type { RandomBeacon } from "../../typechain"
+import type { RandomBeaconStub, RandomBeacon } from "../../typechain"
```
[the remark this answers is not in the document]
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/utils/groups.ts:6]: Addressed in https://github.com/keep-network/keep-core/pull/2671/commits/cbecab1b310bb2bafcdbe474a736e7ea4c505dba being part of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts (hunk 18 above)
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts:296]: Addressed in https://github.com/keep-network/keep-core/pull/2671/commits/e6ac2c4008995b5fb800737735d03c55061c3d03 being part of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts (hunk 19 above)
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts:293]: Addressed in https://github.com/keep-network/keep-core/pull/2671/commits/e6ac2c4008995b5fb800737735d03c55061c3d03 being part of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts (hunk 21 above)
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts:449]: Addressed in https://github.com/keep-network/keep-core/pull/2671/commits/e6ac2c4008995b5fb800737735d03c55061c3d03 being part of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts (hunk 22 above)
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts:468]: Ok, got it!

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts (hunk 23 above)
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts:489]: Addressed in https://github.com/keep-network/keep-core/pull/2671/commits/e6ac2c4008995b5fb800737735d03c55061c3d03 being part of the follow-up PR: https://github.com/keep-network/keep-core/pull/2671

File: solidity/random-beacon/test/RandomBeacon.Relay.test.ts (hunk 24 above)
Comment (lukasz-zimnoch) [reply solidity/random-beacon/test/RandomBeacon.Relay.test.ts:508]: Not yet. I left a TODO here. We can uncomment once group termination is implemented.

**Comment by lukasz-zimnoch** [comment]
@pdyraga All outstanding comments addressed in the follow-up PR: https://github.com/keep-network/keep-core/pull/2671
